app/org/sagebionetworks/bridge/services/backfill/UploadCompletedBackfill.java

```diff
@@ -0,0 +1,111 @@
+package org.sagebionetworks.bridge.services.backfill;
+
+import java.io.IOException;
+import java.util.List;
+import javax.annotation.Resource;
+
+import com.amazonaws.services.s3.AmazonS3;
+import com.amazonaws.services.s3.iterable.S3Objects;
+import com.amazonaws.services.s3.model.S3ObjectSummary;
+
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.stereotype.Component;
+
+import com.google.common.util.concurrent.RateLimiter;
+
+import org.sagebionetworks.bridge.config.BridgeConfig;
+import org.sagebionetworks.bridge.config.BridgeConfigFactory;
+import org.sagebionetworks.bridge.dao.HealthCodeDao;
+import org.sagebionetworks.bridge.models.backfill.BackfillTask;
+import org.sagebionetworks.bridge.models.studies.StudyIdentifierImpl;
+import org.sagebionetworks.bridge.models.upload.Upload;
+import org.sagebionetworks.bridge.models.upload.UploadCompletionClient;
+import org.sagebionetworks.bridge.s3.S3Helper;
+import org.sagebionetworks.bridge.services.UploadService;
+
+/**
+ * Created by liujoshua on 7/27/16.
+ */
+@Component
+public class UploadCompletedBackfill extends AsyncBackfillTemplate {
+    private static final int LOCK_EXPIRY_IN_SECONDS = 60 * 10;
+
+    private static final double UPLOAD_CHECKS_PER_SECOND = 10.0;
+    private static final double UPLOAD_COMPLETIONS_PER_SECOND = 2.0;
+
+    private static final String CONFIG_KEY_UPLOAD_BUCKET = "upload.bucket";
+    private static final String UPLOAD_PREFIX_BUCKET =
+            "org-sagebridge-backfill-" + BridgeConfigFactory.getConfig().getEnvironment().name().toLowerCase();
+    private static final String UPLOAD_PREFIX_FILENAME = "upload-validation-backfill-uploadPrefixes";
+
+    private RateLimiter checkRateLimiter = RateLimiter.create(UPLOAD_CHECKS_PER_SECOND);
+    private RateLimiter completionRateLimiter = RateLimiter.create(UPLOAD_COMPLETIONS_PER_SECOND);
+
+    private UploadService uploadService;
+    private HealthCodeDao healthCodeDao;
+    private AmazonS3 s3Client;
+    private S3Helper s3Helper;
+    private String uploadBucket;
+
+    @Autowired
+    final void setConfig(BridgeConfig config) {
+        uploadBucket = config.getProperty(CONFIG_KEY_UPLOAD_BUCKET);
+    }
+
+    @Resource(name = "s3Client")
+    public void setS3Client(AmazonS3 s3Client) {
```

Comment:
You'll want to put "final" on your Spring setters, to make sure subclasses don't override them and cause mysterious failures later down the line.